relay-sampling/src/lib.rs

```diff
@@ -674,39 +674,84 @@ impl SamplingConfig {
 }
 
 /// The User related information in the trace context
-#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+///
+/// This is more of a mixin to be used in the TraceContext
+#[derive(Debug, Clone, Serialize, Default)]
 pub struct TraceUserContext {
     #[serde(default, skip_serializing_if = "String::is_empty")]
-    pub segment: String,
+    pub user_segment: String,
     #[serde(default, skip_serializing_if = "String::is_empty")]
-    pub id: String,
+    pub user_id: String,
+}
+
+impl<'de> Deserialize<'de> for TraceUserContext {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize, Default)]
+        struct Nested {
+            #[serde(default)]
+            pub segment: String,
+            #[serde(default)]
+            pub id: String,
+        }
+
+        #[derive(Deserialize)]
+        struct Helper {
+            // Nested implements default, but we used to accept user=null (not sure if any SDK
+            // sends this though)
+            #[serde(default)]
+            user: Option<Nested>,
+            #[serde(default)]
+            user_segment: String,
+            #[serde(default)]
+            user_id: String,
+        }
+
+        let helper = Helper::deserialize(deserializer)?;
+
+        if helper.user_id.is_empty() && helper.user_segment.is_empty() {
+            let user = helper.user.unwrap_or_default();
```

Comment:
This is falling back to the nested implementation if _both_ `user_id` and `user_segment` are empty. If I understand and remember correctly, the flat option is the default one (and not the nested one); and I'm making that assumption in the following suggestion: deserialization should error when the trace context contains at least one of the default values, without even checking the nested structure.

Not a blocker though.

Reply:
I also miss some tests with the nested protocol, and with both protocols at the same time.

Reply:
resolved offline, added one more test